src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Folding/FoldingRangeEndpoint.cs

```diff
@@ -136,7 +140,50 @@ public FoldingRangeRegistrationOptions GetRegistrationOptions(FoldingRangeCapabi
                 mappedRanges.AddRange(ranges);
             }
 
-            return new Container<FoldingRange>(mappedRanges);
+            var finalRanges = FinalizeFoldingRanges(mappedRanges, codeDocument);
+            return new Container<FoldingRange>(finalRanges);
+        }
+
+        private IEnumerable<FoldingRange> FinalizeFoldingRanges(List<FoldingRange> mappedRanges, RazorCodeDocument codeDocument)
+        {
+            // Don't allow ranges to be reported if they aren't spanning at least one line
+            var validRanges = mappedRanges.Where(r => r.StartLine < r.EndLine);
+
+            // Reduce ranges that have the same start line to be a single instance with the largest
```

Comment:
Great job on the comments, very helpful for those of us with less context in this area.